Frameworks/CoreGraphics/CGPath.mm

```diff
@@ -35,6 +35,78 @@
 
 static const wchar_t* TAG = L"CGPath";
 
+using namespace Microsoft::WRL;
+
+class _CGPathCustomSink : public RuntimeClass<RuntimeClassFlags<RuntimeClassType::WinRtClassicComMix>, ID2D1SimplifiedGeometrySink> {
+protected:
+    InspectableClass(L"Windows.Bridge.Direct2D._CGPathCustomSink", TrustLevel::BaseTrust);
+
+public:
+    ID2D1GeometrySink* getBackingSink() {
+        return geometrySink.Get();
+    }
+
+    _CGPathCustomSink(_In_ ID2D1GeometrySink* sink) : geometrySink(sink), isFigureOpen(false) {
+    }
+
+    STDMETHOD_(void, _EndFigure)(D2D1_FIGURE_END figureEnd) {
+        if (isFigureOpen) {
+            geometrySink->EndFigure(figureEnd);
+            isFigureOpen = false;
+        }
+    }
+
+    STDMETHOD_(void, SetFillMode)(D2D1_FILL_MODE fillMode) {
+        geometrySink->SetFillMode(fillMode);
+    }
+
+    STDMETHOD_(void, SetSegmentFlags)(D2D1_PATH_SEGMENT vertexFlags) {
+        geometrySink->SetSegmentFlags(vertexFlags);
+    }
+
+    STDMETHOD_(void, AddLines)(_In_reads_(pointsCount) CONST D2D1_POINT_2F* points, UINT32 pointsCount) {
+        geometrySink->AddLines(points, pointsCount);
+        lastPoint = points[pointsCount - 1];
+    }
+
+    STDMETHOD_(void, AddBeziers)(_In_reads_(beziersCount) CONST D2D1_BEZIER_SEGMENT* beziers, UINT32 beziersCount) {
+        geometrySink->AddBeziers(beziers, beziersCount);
+        lastPoint = beziers[beziersCount - 1].point3;
+    }
+
+    STDMETHOD_(void, BeginFigure)(D2D1_POINT_2F startPoint, D2D1_FIGURE_BEGIN figureBegin) {
+        if (isFigureOpen) {
+            if (startPoint.x != lastPoint.x || startPoint.y != lastPoint.y) {
+                _EndFigure(D2D1_FIGURE_END_OPEN);
+                geometrySink->BeginFigure(startPoint, figureBegin);
+            }
+        } else {
+            geometrySink->BeginFigure(startPoint, figureBegin);
+        }
+        isFigureOpen = true;
+    }
+
+    STDMETHOD_(void, EndFigure)(D2D1_FIGURE_END figureEnd) {
+    }
+
+    STDMETHOD(Close)() {
+        return S_OK;
+    };
+
+    STDMETHOD(_Close)() {
```

Comment:
We don't want the simplify method from D2D to call close and invalidate our figure work. We want full control over closing the geometrysink. If close is called, it may put our geometrysink in a bad state and prevent further calls.

---
In reply to: [96963185](https://github.com/Microsoft/WinObjC/pull/1738#discussion_r96963185) [](ancestors = 96963185)